src/System.Security.Cryptography.ProtectedData/tests/ProtectedDataTests.cs

```diff
@@ -0,0 +1,139 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Security.Cryptography;
+
+using Test.Cryptography;
+using Xunit;
+
+using DATA_BLOB = Interop.Crypt32.DATA_BLOB;
+using CryptProtectDataFlags = Interop.Crypt32.CryptProtectDataFlags;
+
+namespace System.Security.Cryptography.ProtectedDataTests
+{
+    public static class ProtectedDataTests
+    {
+        [Fact]
+        public static void RoundTripWithoutEntropyCurrentUser()
+        {
+            byte[] plain = { 1, 2, 3 };
+            RoundTrip(plain, null, DataProtectionScope.CurrentUser);
+        }
+
+        [Fact]
+        public static void RoundTripWithoutEntropyLocalMachine()
+        {
+            byte[] plain = { 1, 2, 3 };
+            RoundTrip(plain, null, DataProtectionScope.LocalMachine);
+        }
+
+        [Fact]
+        public static void RoundTripWithEntropyCurrentUser()
+        {
+            byte[] plain = { 1, 2, 3 };
+            byte[] entropy = { 4, 5, 6 };
+            RoundTrip(plain, null, DataProtectionScope.CurrentUser);
+        }
+
+        [Fact]
+        public static void RoundTripWithEntropyLocalMachine()
+        {
+            byte[] plain = { 1, 2, 3 };
+            byte[] entropy = { 4, 5, 6 };
+            RoundTrip(plain, null, DataProtectionScope.LocalMachine);
+        }
+
+        private static void RoundTrip(byte[] plain, byte[] optionalEntropy, DataProtectionScope scope)
+        {
+            {
+                // Protect using CoreFx, unprotect using direct p/invoke.
+                byte[] encrypted = ProtectedData.Protect(plain, optionalEntropy, scope);
+                byte[] recovered = UnprotectUsingPInvoke(encrypted, optionalEntropy, scope);
+                Assert.Equal<byte>(plain, recovered);
+            }
+
+            {
+                // Protect using P/Invoke, unprotect using CoreFx.
+                byte[] encrypted = ProtectUsingPInvoke(plain, optionalEntropy, scope);
+                byte[] recovered = ProtectedData.Unprotect(encrypted, optionalEntropy, scope);
+                Assert.Equal<byte>(plain, recovered);
+            }
+        }
+
+
+        private static byte[] ProtectUsingPInvoke(byte[] inputData, byte[] optionalEntropy, DataProtectionScope scope)
+        {
+            return ProtectOrUnprotect(inputData, optionalEntropy, scope, protect: true);
+        }
+
+        private static byte[] UnprotectUsingPInvoke(byte[] inputData, byte[] optionalEntropy, DataProtectionScope scope)
+        {
+            return ProtectOrUnprotect(inputData, optionalEntropy, scope, protect: false);
+        }
+
+        private static byte[] ProtectOrUnprotect(byte[] inputData, byte[] optionalEntropy, DataProtectionScope scope, bool protect)
```

Comment:
In an ideal world, this would be a clean-room implementation by a different person. But that’s what happens when we eliminate testers.